### ISSUE TITLE
merge-test-replace script to use instead of continuous integration

### ISSUE THREAD

**Issue by andres-erbsen** [mit-plv/fiat-crypto#447]
I think the flow should be roughly as follows:

- [ ] script to make sure master is up to date, check out a new copy of master, merge a new branch into the copy of master
- [x] remote async test runner: the script I wrote for running tests on CSAIL SLURM works for me: https://gist.github.com/andres-erbsen/9b0e79eb49acf382ecda479023abcd41
In particular, it survives network partitions, even in the case when tests finish during a network partition, and builds all of fiat-crypto with coq master, 8.9, and 8.8.
- [ ] script to commit the merge, move put the merge commit on top of master, and push to github. Ideally it would always leave local master and remote master in a consistent state. **What should it do in case somebody else has updated master in the meantime?** As we currently test before merging, the most similar thing to do would be to just rebase/merge that commit if it can be done automatically. Of course this doesn't guarantee that master always builds.
- [ ] glue these three scripts together

- ( ) fix or drop build with coq 8.7 (coqdep being slow?)
- ( ) local multi-version test runner: we should also probably make a local version of this script for those without access to MIT SLURM. The current script relies on SLURM infrastructure for limiting concurrency, so this is not totally trivial.
- ( ) out-of-source build (nice-to-have): build with multiple coq versions without copying the source directory?
Currently 8

**Comment by andres-erbsen** [comment]
@JasonGross and I discussed this and decided that the flow should be as follows

0) the code to be merged is already checked into a copy of the correct repository
1) fetch latest master if possible
2) create new folder for master from (1), merge the feature branch (0) into that
3) test, either locally or something like `test-remotely.sh`. Abort if tests fail.
4) try to push the (2) to upstream

If the push fails because the the branch from (2) is not up to date, merge anyway as follows

5) fetch latest master
6) merge feature branch (0) into master from (5), noting in the merge commit message the commit hash of master (1) and that tests passed on top of that
7) try to push the merged version (6) to upstrea, retry 5,6,7 a couple of times if that fails.

Ideally, the merge commits would also record the build time diff. As times only make sense if they are measured on the same machine, this would require testing both before and after the merge (probably in parallel). Currently @JasonGross has been storing the diffs in commit messages, but that is not ideal as it makes the commit log harder to read. Alternatively the logs could be stored in the repositroy, but it is unclear what a good format would be.

**Comment by JasonGross** [comment]
> but it is unclear what a good format would be

I see two options:
1. The output of `make --output-sync TIMED=1` (possibly after passing it through `grep 'real:' to eliminate non-timing lines).  This can then be passed into either `./etc/coq-scripts/timing/make-one-time-file.py` for a sorted table of the current commit, or two of them can be passed into `./etc/coq-scripts/timing/make-both-time-files.py`.
2. The table of the output of `./etc/coq-scripts/timing/make-both-time-files.py` (possibly also `./etc/coq-scripts/timing/make-one-time-file.py`).

Note that (1) has the advantage that we can later build tables that compare across multiple commits.  If desired, after grepping for `real:`, we can pass them through `sed s'/^\(.*real: \)\([^, ]*\)\(.*\)$/\2 \1\3/g' | sort -gr | sed s'/^\([^ ]*\) \(.*real: \)\(.*\)$/\2\1\3/g'`, which will give us a sorted record of the time for the current commit, and will still serve as valid input to the table-printer.

**Comment by andres-erbsen** [comment]
I am fine with both of the above. However, that is not really what is puzzling me. I am wondering how to sort measurements from different machines and from different commits, and to limit commit chaff while doing so.

**Comment by JasonGross** [comment]
My first sense: for each commit/branch that we do timing for, we pick a single version of Coq and a single machine (slurm?), and store in the merge commit the timing of the master build we're merging into, and the timing of the merged branch.  If we want, we can also store the table of the diff.  These files will basically be entirely rewritten on every merge commit created by this script.

**Comment by JasonGross** [comment]
Current error message:
```
$ ../test-remotely/test-remotely.sh coq standalone c-files new-pipeline TIMED=1
+ klist -l
+ grep -m1 -i -o -e ^\S\+@csail.mit.edu\b
+ cut -d@ -f1
+ krbuser=jgross
+ [ -z jgross ]
+ destination=jgross@tig-slurm.csail.mit.edu
+ pwd
+ basename /home/jgross/Documents/repos/fiat-crypto-stage-new
+ rdir_template=/data/scratch/jgross/fiat-crypto-stage-new-XXXXXXXXXXXXX
+ remote mktemp -d /data/scratch/jgross/fiat-crypto-stage-new-XXXXXXXXXXXXX
+ ssh -K jgross@tig-slurm.csail.mit.edu -- mktemp -d /data/scratch/jgross/fiat-crypto-stage-new-XXXXXXXXXXXXX
+ rdir=/data/scratch/jgross/fiat-crypto-stage-new-06XDXM1IsWYOq
+ tar --exclude-vcs-ignores --exclude-vcs -c -f - .
+ remote dd bs=1M of=/data/scratch/jgross/fiat-crypto-stage-new-06XDXM1IsWYOq/src.tar
+ ssh -K jgross@tig-slurm.csail.mit.edu -- dd bs=1M of=/data/scratch/jgross/fiat-crypto-stage-new-06XDXM1IsWYOq/src.tar
0+3601 records in
0+3601 records out
48885760 bytes (49 MB, 47 MiB) copied, 0.855114 s, 57.2 MB/s
+ dd if=/dev/urandom bs=16 count=1
+ xxd -p
+ needle=3ca4c9bff8c3fb5b96fc3b584e57192c
+ mktemp -u
+ moshout=/tmp/tmp.EELMf589wg
+ mkfifo -m 600 /tmp/tmp.EELMf589wg
+ coqversions=master v8.9 v8.8
+ moshpid=28761
+ tee /tmp/tmp.EELMf589wg
+ cat
+ mosh --ssh ssh -K jgross@tig-slurm.csail.mit.edu --no-init --predict=never -- sh -euc+    rdir="$1"; shift;
  needle="$1"; shift;
  coqversions="master v8.9 v8.8";

  pids=""
  for coqversion in $coqversions; do
    dir="$rdir/$coqversion";
    mkdir "$dir";
    touch "$dir/build.log";
    (
      tar -C "$dir" -x -f "$rdir/src.tar";
      export PATH="/data/scratch/andreser/coq/$coqversion/bin:/data/scratch/andreser/coq/common/bin:$PATH";
      exec salloc -p cpu -N2-4 /usr/bin/time make --output-sync COQC="srun -N1 -n1 --exclusive coqc" -j -C "$rdir/$coqversion" "$@"
    ) > "$dir/build.log" 2>&1 &
    pids="$pids $!"
  done;

  printf "shellpid: %s\ncoqversions: %s\npids: %s\n" "$$" "$coqversions" "$pids" | tr ' ' '\t' | column -t | tee "$rdir/started";

  ( cd "$rdir" ; exec tail -f -- */build.log ) &
  tailpid=$!;

  exitcodes="";
  code=0;
  for pid in $pids; do
    pidcode=0;
    wait "$pid" || pidcode="$?"
    exitcodes="$exitcodes $pidcode"
    if [ "$pidcode" -gt "$code" ]; then
      code="$pidcode";
    fi
  done;

  for coqversion in $coqversions; do printf "" >> "$rdir/$coqversion/build.log"; done;
  sleep 1;
  kill "$tailpid";

  printf "================================\n" ;
  printf "coqversions: %s\nexitcodes: %s\nexitcode=%s=%s\n" \
    "$coqversions" "$exitcodes" "$code" "$needle" \
    | tr ' ' '\t' | column -t | tee "$rdir/finished";

  exec sleep "$((365*24*60*60))"; # hang to WORKAROUND(https://github.com/mobile-shell/mosh/issues/1011)grep test-remotely-shell /data/scratch/jgross/fiat-crypto-stage-new-06XDXM1IsWYOq -m1 3ca4c9bff8c3fb5b96fc3b584e57192c -o coq exitcode=[0-9]*=3ca4c9bff8c3fb5b96fc3b584e57192c standalone+
 c-files new-pipeline TIMED=1
cut -d= -f2
Server started without pseudo-terminal. Opening 80x24 terminal.

tcgetattr: Inappropriate ioctl for device
+ rexit=
+ kill 28761
../test-remotely/test-remotely.sh: 71: kill: No such process
```

Relevant bit is probably `tcgetattr: Inappropriate ioctl for device`, which only showed up after I removed the `>/dev/null` from the script.
